File: src/main/java/org/agentic/flink/example/markets/producer/BondSecuritiesProducer.java
package org.agentic.flink.example.markets.producer;

import org.agentic.flink.example.markets.model.MarketRecords.Security;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Java twin of {@code examples-bin/markets/bond_securities_producer.py}. Publishes a slow stream of
 * security-master rows to the {@code fnd-securities} topic. The Flink job broadcasts these into
 * the inventory enrichment stage.
 *
 * <pre>
 *   java -cp target/agentic-flink-1.0.0-SNAPSHOT.jar \
 *     org.agentic.flink.example.markets.producer.BondSecuritiesProducer --count 50000 --rate 500
 * </pre>
 */
public final class BondSecuritiesProducer {

  private static final List<String> SECTORS =
      List.of("Financials", "Industrials", "Utilities", "Energy", "Healthcare", "Tech", "Consumer");
  private static final Map<String, List<String>> INDUSTRIES =
      Map.of(
          "Financials", List.of("Banks", "Insurance", "AssetMgmt"),
          "Industrials", List.of("Machinery", "Aerospace", "Logistics"),
          "Utilities", List.of("Electric", "Gas", "Water"),
          "Energy", List.of("Oil", "Renewables"),
          "Healthcare", List.of("Pharma", "Devices"),
          "Tech", List.of("Software", "Semiconductors"),
          "Consumer", List.of("Retail", "Apparel"));
  private static final List<String> FITCH =
      List.of("AAA", "AA+", "AA", "AA-", "A+", "A", "A-", "BBB+", "BBB", "BBB-", "BB+", "BB", "NR");
  private static final List<String> MOODY =
      List.of("Aaa", "Aa1", "Aa2", "A1", "A2", "A3", "Baa1", "Baa2", "Baa3", "Ba1", "Ba2", "NR");
  private static final java.util.Set<String> IG =
      java.util.Set.of("AAA", "AA+", "AA", "AA-", "A+", "A", "A-", "BBB+", "BBB", "BBB-");

  public static void main(String[] args) throws Exception {
    int count = BondInventoryProducer.parseInt(args, "--count", 50_000);
    int rate = BondInventoryProducer.parseInt(args, "--rate", 500);
    long sleepNanos = Math.max(1_000_000L, 1_000_000_000L / Math.max(rate, 1));

    try (MarketProducerSupport mp =
        new MarketProducerSupport(MarketProducerSupport.defaultBootstrap(), "bond-securities")) {
      System.out.printf(Locale.ROOT,
          "producing %d securities to fnd-securities at ~%d/s%n", count, rate);
      for (int i = 1; i <= count && !Thread.currentThread().isInterrupted(); i++) {
        mp.send("fnd-securities", randomSecurity(i));
        java.util.concurrent.locks.LockSupport.parkNanos(sleepNanos);
      }
    }
  }

  static Security randomSecurity(int id) {
    ThreadLocalRandom rng = ThreadLocalRandom.current();
    String sector = SECTORS.get(rng.nextInt(SECTORS.size()));
    String fitch = FITCH.get(rng.nextInt(FITCH.size()));
    String isin = String.format(Locale.ROOT, "FN%010d", id);
    String maturity =
        String.format(
            Locale.ROOT, "203%d-%02d-%02d",
            rng.nextInt(0, 6), 1 + rng.nextInt(9), 10 + rng.nextInt(18));
    return new Security(
        id, isin, String.format(Locale.ROOT, "%09d", id),
        String.format(Locale.ROOT, "FNDP_%05d", id),
        String.format(Locale.ROOT, "FNDP_%05d", id),
        sector, INDUSTRIES.get(sector).get(rng.nextInt(INDUSTRIES.get(sector).size())),
        Math.round(rng.nextDouble(0.005, 0.085) * 10_000.0) / 10_000.0,
        maturity, fitch, FITCH.get(rng.nextInt(FITCH.size())),
        MOODY.get(rng.nextInt(MOODY.size())),
        IG.contains(fitch) ? "Y" : "N");
  }
}

File: examples-bin/markets/bond_securities_producer.py
#!/usr/bin/env python3
"""Anonymised security-master producer (issuer, sector, ratings) for the bond markets example.

Writes a slow, low-rate stream of records to ``fnd-securities`` — the Flink job broadcasts these
into the inventory enrichment stage. Fields mirror the user's mrkaxis ``kafka_security_producer.py``
schema with anonymised issuer/sector names.

    pip install kafka-python numpy
    python examples-bin/markets/bond_securities_producer.py
"""

from __future__ import annotations

import argparse
import time
from datetime import datetime, timezone

import numpy as np
from kafka import KafkaProducer  # type: ignore

import kafka_config

rng = np.random.default_rng(2026)

SECTORS = ["Financials", "Industrials", "Utilities", "Energy", "Healthcare", "Tech", "Consumer"]
INDUSTRIES = {
    "Financials": ["Banks", "Insurance", "AssetMgmt"],
    "Industrials": ["Machinery", "Aerospace", "Logistics"],
    "Utilities": ["Electric", "Gas", "Water"],
    "Energy": ["Oil", "Renewables"],
    "Healthcare": ["Pharma", "Devices"],
    "Tech": ["Software", "Semiconductors"],
    "Consumer": ["Retail", "Apparel"],
}
FITCH = ["AAA", "AA+", "AA", "AA-", "A+", "A", "A-", "BBB+", "BBB", "BBB-", "BB+", "BB", "NR"]
SNP = FITCH
MOODY = ["Aaa", "Aa1", "Aa2", "A1", "A2", "A3", "Baa1", "Baa2", "Baa3", "Ba1", "Ba2", "NR"]
COUNTRIES = ["US", "UK", "DE", "FR", "CA"]
CCY = ["USD", "EUR", "GBP", "CHF"]


def gen_security(idx: int) -> dict:
    sec = rng.choice(SECTORS)
    rating = rng.choice(FITCH)
    investment_grade = "Y" if rating in {"AAA", "AA+", "AA", "AA-", "A+", "A", "A-", "BBB+", "BBB", "BBB-"} else "N"
    return dict(
        id=int(idx),
        isin=f"FN{idx:010d}",
        cusip=f"{idx:09d}",
        name=f"FNDP_{idx:05d}",
        issuer=f"FNDP_{idx:05d}",
        sector=str(sec),
        industry=str(rng.choice(INDUSTRIES[sec])),
        coupon=round(float(rng.uniform(0.005, 0.085)), 4),
        maturity=f"203{rng.integers(0, 6)}-0{rng.integers(1, 10)}-{rng.integers(10, 28):02d}",
        fitchRating=str(rating),
        snpRating=str(rng.choice(SNP)),
        moodyRating=str(rng.choice(MOODY)),
        isInvestmentGrade=investment_grade,
        callable_="Y" if rng.random() < 0.3 else "N",
        paymentFrequency=str(rng.choice(["Q", "S", "A"])),
        countryCode=str(rng.choice(COUNTRIES)),
        currency=str(rng.choice(CCY)),
        timestamp=datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
    )


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--count", type=int, default=50_000, help="number of securities to publish")
    ap.add_argument("--rate", type=float, default=500, help="rows per second")
    args = ap.parse_args()

    producer = KafkaProducer(**kafka_config.producer_kwargs())
    topic = kafka_config.TOPIC_BOND_SECURITIES
    interval = 1.0 / max(args.rate, 1)
    print(f"producing {args.count} securities to '{topic}' at ~{args.rate}/s")
    try:
        for i in range(1, args.count + 1):
            producer.send(topic, value=gen_security(i))
            if i % 1000 == 0:
                print(f"  sent {i} securities")
            time.sleep(interval)
    except KeyboardInterrupt:
        print("\nstopping")
    finally:
        producer.flush()
        producer.close()


if __name__ == "__main__":
    main()
